source/s3_checksum_stream.c

```diff
@@ -0,0 +1,122 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include "aws/s3/private/s3_checksums.h"
+#include <aws/common/encoding.h>
+#include <aws/io/stream.h>
+
+struct aws_checksum_stream {
+    struct aws_input_stream *old_stream;
+    struct aws_s3_checksum *checksum;
+    struct aws_byte_buf checksum_result;
+    /* base64 encoded checksum of the stream, updated on destruction of sream */
+    struct aws_byte_buf *encoded_checksum_output;
+    bool did_seek;
+};
+
+static int s_aws_input_checksum_stream_seek(
+    struct aws_input_stream *stream,
+    int64_t offset,
+    enum aws_stream_seek_basis basis) {
+    (void)stream;
+    (void)offset;
+    (void)basis;
+    AWS_LOGF_ERROR(
+        AWS_LS_S3_CLIENT,
+        "Cannot seek on checksum stream, as it will cause the checksum output to mismatch the checksum of the stream "
+        "contents");
+    AWS_ASSERT(false);
+    return aws_raise_error(AWS_IO_STREAM_INVALID_SEEK_POSITION);
+}
+
+static int s_aws_input_checksum_stream_read(struct aws_input_stream *stream, struct aws_byte_buf *dest) {
+    struct aws_checksum_stream *impl = stream->impl;
+
+    size_t start = dest->len;
+    int err = aws_input_stream_read(impl->old_stream, dest);
+    size_t end = dest->len;
+    struct aws_byte_cursor to_sum = aws_byte_cursor_from_buf(dest);
+    to_sum.ptr += start;
+    to_sum.len = end - start;
+    if (!err) {
+        int checksum_res = aws_checksum_update(impl->checksum, &to_sum);
+        if (checksum_res) {
+            dest->len = start;
+            impl->did_seek = true;
+        }
+        return checksum_res;
+    }
+    return err;
+}
+
+static int s_aws_input_checksum_stream_get_status(struct aws_input_stream *stream, struct aws_stream_status *status) {
+    struct aws_checksum_stream *impl = stream->impl;
+    return aws_input_stream_get_status(impl->old_stream, status);
+}
+
+static int s_aws_input_checksum_stream_get_length(struct aws_input_stream *stream, int64_t *out_length) {
+    struct aws_checksum_stream *impl = stream->impl;
+    return aws_input_stream_get_length(impl->old_stream, out_length);
+}
+
+/* We take ownership of the old inputstream, and destroy it with this input stream. This is because we want to be able
+ * to substitute in the chunk_stream for the cursor stream currently used in s_s3_meta_request_default_prepare_request
+ * which returns the new stream. So in order to prevent the need of keeping track of two input streams we instead
+ * consume the cursor stream and destroy it with this one */
+static void s_aws_input_checksum_stream_destroy(struct aws_input_stream *stream) {
+    if (stream) {
+        struct aws_checksum_stream *impl = stream->impl;
+        AWS_FATAL_ASSERT(aws_checksum_finalize(impl->checksum, &impl->checksum_result, 0) == AWS_OP_SUCCESS);
```

Comment:
Instead of FATAL ASSERT, maybe just ASSERT the return code and if error happens, set the `checksum_result` to empty. So that, the chunked_stream can check the empty checksum and return error properly from https://github.com/awslabs/aws-c-s3/pull/181/files#diff-4be41bb6016233fcddc834d9baff1a97b14989916f4323e3d7ca167e05f37a8aR60-R62